can/interfaces/vector/canlib.py

```diff
@@ -474,65 +475,101 @@ def handle_canfd_event(self, event: xlclass.XLcanRxEvent) -> None:
         """Handle non-message CAN FD events.
 
         Method is called by :meth:`~can.interfaces.vector.VectorBus._recv_internal`
-        when `event.tag` is not `XL_RECEIVE_MSG`. Subclasses can implement this method.
+        when `event.tag` is not `XL_CAN_EV_TAG_RX_OK` or `XL_CAN_EV_TAG_TX_OK`.
+        Subclasses can implement this method.
 
-        :param event: `XLcanRxEvent` that could have a `XL_CAN_EV_TAG_RX_ERROR`, `XL_CAN_EV_TAG_TX_ERROR`
-            or `XL_CAN_EV_TAG_CHIP_STATE` tag.
+        :param event: `XLcanRxEvent` that could have a `XL_CAN_EV_TAG_RX_ERROR`, `XL_CAN_EV_TAG_TX_ERROR`,
+            `XL_TIMER` or `XL_CAN_EV_TAG_CHIP_STATE` tag.
         :return: None
         """
         pass
 
-    def send(self, msg, timeout=None):
-        msg_id = msg.arbitration_id
+    def send(self, msg: Message, timeout: typing.Optional[float] = None):
+        self._send_sequence([msg])
+
+    def _send_sequence(self, msgs: typing.Sequence[Message]) -> int:
+        """Send messages and return number of successful transmissions."""
+        if self.fd:
+            return self._send_can_fd_msg_sequence(msgs)
+        else:
+            return self._send_can_msg_sequence(msgs)
+
+    def _get_tx_channel_mask(self, msgs: typing.Sequence[Message]) -> int:
+        if len(msgs) == 1:
+            return self.channel_masks.get(msgs[0].channel, self.mask)
+        else:
+            return self.mask
+
+    def _send_can_msg_sequence(self, msgs: typing.Sequence[Message]) -> int:
+        """Send CAN messages and return number of successful transmissions."""
+        mask = self._get_tx_channel_mask(msgs)
+        message_count = ctypes.c_uint(len(msgs))
+
+        xl_event_array = (xlclass.XLevent * message_count.value)(
+            *map(self._build_xl_event, msgs)
+        )
+
+        xldriver.xlCanTransmit(self.port_handle, mask, message_count, xl_event_array)
+        return message_count.value
 
+    @staticmethod
+    def _build_xl_event(msg: Message) -> xlclass.XLevent:
+        msg_id = msg.arbitration_id
         if msg.is_extended_id:
             msg_id |= xldefine.XL_MessageFlagsExtended.XL_CAN_EXT_MSG_ID.value
 
         flags = 0
+        if msg.is_remote_frame:
+            flags |= xldefine.XL_MessageFlags.XL_CAN_MSG_FLAG_REMOTE_FRAME.value
 
-        # If channel has been specified, try to send only to that one.
-        # Otherwise send to all channels
-        mask = self.channel_masks.get(msg.channel, self.mask)
+        xl_event = xlclass.XLevent()
+        xl_event.tag = xldefine.XL_EventTags.XL_TRANSMIT_MSG.value
+        xl_event.tagData.msg.id = msg_id
+        xl_event.tagData.msg.dlc = msg.dlc
+        xl_event.tagData.msg.flags = flags
+        xl_event.tagData.msg.data = tuple(msg.data)
```

Comment:
Have you verified that this works for DLC < 8?

Reply:
Just tried it, no problems here.